Bottom: bd775c9e72e1749321c3149b84f9bb80420d03de
Top:    b8386f53d00f8b85c6231e904e3724a50703b000
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2009-11-04 01:22:26 +0000

Refresh of tag-circular-contours-so-they-

---

diff --git a/src/hid/common/hidgl.c b/src/hid/common/hidgl.c
index 54c2b3b..a582d3d 100644
--- a/src/hid/common/hidgl.c
+++ b/src/hid/common/hidgl.c
@@ -602,10 +602,26 @@ do_hole (const BoxType *b, void *cl)
 {
   struct do_hole_info *info = cl;
   PLINE *curc = (PLINE *) b;
+  cairo_traps_t *traps;
+
   /* Ignore the outer contour - we draw it first explicitly*/
   if (curc->Flags.orient == PLF_DIR) {
     return 0;
   }
+
+  /* If the contour is round, and hidgl_fill_circle would use
+   * less slices than we have vertices to draw it, then call
+   * hidgl_fill_circle to draw this contour.
+   */
+  if (curc->is_round) {
+    double slices = M_PI * 2 * curc->radius /
+                    info->scale / PIXELS_PER_CIRCLINE;
+    if (slices < curc->Count) {
+      hidgl_fill_circle (curc->cx, curc->cy, curc->radius, info->scale);
+      return 1;
+    }
+  }
+
   tesselate_contour (info->tobj, &curc->head, info->vertices);
   return 1;
 }
